### PR TITLE
SIMD printing with modcc.

### PR DESCRIPTION
**Pull request by halfflat**
Merges master branch.

Config/build changes:
* Split arbor modcc vectorization option (now `ARB_VECTORIZE`) and target-architecture optimization (now `ARB_ARCH`).
* For `arbor` and `arbormech` targets, and in particular not the `modcc` target, use `ARB_ARCH` to generate corresponding target-appropriate binaries, including, for example, appropriate SIMD support.
* Extend `CompilerOptions.cmake` to map as best as able between the various target architecture names (we use the gcc names) and the correct option to pass to the compiler based on the compiler and platform.
* Add work-around for misidentification by CMake of XL C as Clang.
* As a temporary work-around, include `arbormech` library twice on link line to resolve circular arbor–arbormech dependencies.

SIMD class changes:
* Incorporate SIMD module printing into modcc cprinter.
* Add `simd_cast` operation between SIMD value types of the same width, and with `std::array`. (Note: this was tested and used in an early development version of the code, but not in this version. It was still a lacuna in the original SIMD wrappers, so it has been left in.)
* Restructure SIMD gather/scatter API to use a `simd::indirect` expression, which encapsulates a pointer and SIMD offset.
* Add `simd::index_constraint` scoped enum to describe knowledge of contention in indirect indices, so that we can branch on this to the appropriate implementation.
* Add SIMD concrete implementation routines `reduce_add` for horizontal reduction and `element0` for access to first lane scalar value.
* Add SIMD value method `sum()` that exposes implementation `reduce_add`.
* Add SIMD concrete implementation routine `compound_indexed_add` that provides the implementation for `indirect(p, simd_indices) += simd_value` construction.
* Fix SIMD `implbase` bug where some static methods were using the `implbase` fall-back functions instead of the derived class specialized implementations.
* Move SIMD mathematical functions into friend routines of `simd_impl` in order to resolve implicit conversions from scalars in mixed SIMD-scalar operations.
* Use a templated `tag` class to dispatch on SIMD concrete implementation types, to avoid problems with incomplete types in method signatures.
* Remove old SIMD intrinsics.

FVM class changes:
* Add signed `fvm_index_type` (alias for `int`) for the representation of indices (such as CV index, ion instance index etc.) in fvm layouts and mechanisms. This was chosen as an alternative to making `unsigned` versions of all our SIMD implementation classes.
* Make mechanism `width_` a pointer into `data_` for uniform access in generated modules.

Modcc/module changes:
* Modify SIMD printing options for `modcc`: `-s <simd-abi>` specifies the vectorization target: none, avx, avx2, avx512, default, or native. `-S <N>` allows a specific data width to be prescribed.
* Fix problem in `test_ca.mod` with uninitialized ion current.
* Extend cprinter to incorporate SIMD-specific output.
* Add infrastructure support to allow future pre-computation of SIMD index conflict cases for (hopefully) faster scatters and updates.
* Remove old SIMD printing code.

Unit test changes:
* Extend repertoire of generic sequence equality/near equality testing support in `common.hpp`.
* Add warning suppression for icc for the malloc instrumentation code.
* SIMD unit tests for indirect expressions, compound indirect add, reduction.
* Make some exact tests into floating point 'near' tests when comparing computed areas and lengths in swc and fvm layout tests, to account for compiler (e.g. icc) performing semantically inequivalent floating point operation reordering or fusion at -O3.

TODO:
1. SIMD documentation needs to be updated to describe new API and functionality.
2. Unit tests need to be supplied for new modcc cprinter.
3. Padded index implementation for density channels needs to be re-examined, as currently we cannot take advantage of a faster index constraint than `none`. (I'll detail this in a new ticket.)

### CODE REVIEW

**Comment by halfflat** [comment]
Superb, thank you for the update!

**Comment by bcumming** [comment]
We can defer much of the low level documentation, however the installation guide should be changed to:
- reflect the new vectorization options.
- reflect changes to cross-compilation workflow.

**Comment by bcumming** [comment]
`ARB_ARCH` should be inherited by the unit tests so that they test the architecure for which the library was compiled.

**Comment by bcumming** [comment]
Also, add `mechanisms/generated` to `.gitignore`.

File: cmake/CompilerOptions.cmake [hunk 1]
```diff
@@ -65,3 +76,82 @@ if(${CMAKE_CXX_COMPILER_ID} MATCHES "Intel")
...
+
+        # Use -mcpu for all supported targets _except_ for x86, where it should be -march.
+
+        if(target_model MATCHES "x86" OR target_model MATCHES "amd64")
+            set(CXXOPT_ARCH "-march=${arch}")
+        else()
+            set(CXXOPT_ARCH "-mcpu=${arch}")
+        endif()
+
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Intel")
```
[the remark this answers is not in the document]
Comment (halfflat) [reply cmake/CompilerOptions.cmake]: I'll pass that one on to you!

File: CMakeLists.txt [hunk 2]
```diff
@@ -208,20 +208,18 @@ if(ARB_WITH_PROFILING)
 endif()
 
 #----------------------------------------------------------
-# vectorization target
+# Modcc vectorization target
 #----------------------------------------------------------
-set(ARB_VECTORIZE_TARGET "none" CACHE STRING "CPU target for vectorization {none,KNL,AVX2,AVX512}")
-set_property(CACHE ARB_VECTORIZE_TARGET PROPERTY STRINGS none KNL AVX2 AVX512)
+set(ARB_VECTORIZE "none" CACHE STRING "SIMD ABI for generated code vectorization {none, native, avx, avx2, avx512}")
```
[the remark this answers is not in the document]
Comment (halfflat) [reply CMakeLists.txt]: `ARB_ARCH` and `ARB_VECTORIZE` need to be independent for cross-compilation scenarios. That said, we can make the latter a toggle and have the generated mechanisms use the `simd_abi::native` implementations.

Proposed fix:
* Make `modcc` take a binary option `--simd` that enables the SIMD printing back-end and defaults the SIMD ABI to `simd_abi::native`.
* Rename `--simd-width` to `--simd-abi` and have it take an ABI string (i.e. avx, avx2, native etc.) with optional explicit side suffix, so we can override it in development scenarios.
* Add `ARB_MODCC_FLAGS` option to top-level CMake, defaults to empty.
* Make `ARB_VECTORIZE` a toggle that adds the `--simd` flag to `modcc` invocations.

File: cmake/CompilerOptions.cmake [hunk 3]
```diff
@@ -65,3 +76,82 @@ if(${CMAKE_CXX_COMPILER_ID} MATCHES "Intel")
...
+        # icc 17 recognizes the following specific microarchitecture names for -mtune:
+        #     broadwell, haswell, ivybridge, knl, sandybridge, skylake
+
+        if(arch MATCHES "sandybridge")
+            set(tune "${arch}")
+            set(arch "AVX")
+        elseif(arch MATCHES "ivybridge")
+            set(tune "${arch}")
+            set(arch "CORE-AVX-I")
+        elseif(arch MATCHES "broadwell|haswell|skylake")
```
[the remark this answers is not in the document]
Comment (halfflat) [reply cmake/CompilerOptions.cmake]: Gcc's name for this is `skylake-avx512`. I'd like to postpone adding that to the translation list until we have a machine to play with, but we can roll it into this.

File: modcc/printer/cprinter.cpp [hunk 4]
```diff
@@ -252,20 +418,17 @@ void emit_state_read(std::ostream& out, LocalVariable* local) {
 }
 
 void emit_state_update(std::ostream& out, Symbol* from, IndexedVariable* external) {
-    if (external->is_write()) {
-        const char* op = external->op()==tok::plus? "+=": "-=";
-        out << cprint(external) << op << cprint(from) << ";\n";
-    }
+    if (!external->is_write()) return;
```
[the remark this answers is not in the document]
Comment (halfflat) [reply modcc/printer/cprinter.cpp]: I like the early exits, as it lets the functional code stand on its own. But it's no biggie.

File: src/backends/multicore/mechanism.cpp [hunk 5]
```diff
@@ -31,7 +34,7 @@ namespace multicore {
 // forward iterators.
 
 template <typename Source, typename Dest, typename Fill>
-void copy_extend(const Source& source, Dest& dest, const Fill& fill) {
+void copy_extend(const Source& source, Dest&& dest, const Fill& fill) {
```
[the remark this answers is not in the document]
Comment (halfflat) [reply src/backends/multicore/mechanism.cpp]: `Dest` can be an rvalue non-owning but writeable range, and this picks up that case.

File: src/backends/multicore/shared_state.cpp [hunk 6]
```diff
@@ -87,7 +87,7 @@ void ion_state::nernst(fvm_value_type temperature_K) {
         simd_value_type xi(Xi_.data()+i);
         simd_value_type xo(Xo_.data()+i);
 
-        auto ex = factor*simd::log(xo/xi);
+        auto ex = factor*log(xo/xi);
```
[the remark this answers is not in the document]
Comment (halfflat) [reply src/backends/multicore/shared_state.cpp]: Almost certainly! Will move.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (bcumming) [reply CMakeLists.txt]: Sounds good: you can defer the `ARB_MODCC_FLAGS` to a later date if you want.

**Comment by halfflat** [comment]
Regarding generated mechanisms, they really need to be put in the build directory. (Currently, maintaining multiple build directories requires manually cleaning out the `mechanisms/generated` directory, and it's also just good practice.)

Shall I move them in there in this PR?

**Pull request merged**